lib/spells/effects/Moat.cpp

```diff
@@ -0,0 +1,179 @@
+	handler.serializeBool("trigger", trigger);
+	handler.serializeBool("trap", trap);
+	handler.serializeBool("removeOnTrigger", removeOnTrigger);
+	handler.serializeBool("dispellable", dispellable);
+	handler.serializeInt("moatDamage", moatDamage);
+	serializeMoatHexes(handler, "moatHexes", moatHexes);
+	handler.serializeId("triggerAbility", triggerAbility, SpellID::NONE);
+	handler.serializeStruct("defender", sideOptions); //Moats are defender only
+
+	assert(!handler.saving);
+	{
+		auto guard = handler.enterStruct("bonus");
+		const JsonNode & data = handler.getCurrent();
+
+		for(const auto & p : data.Struct())
+		{
+			//TODO: support JsonSerializeFormat in Bonus
+			auto guard = handler.enterStruct(p.first);
+			const JsonNode & bonusNode = handler.getCurrent();
+			auto b = JsonUtils::parseBonus(bonusNode);
+			bonus.push_back(b);
+		}
+	}
+}
+
+void MoatPlacer::convertBonus(const Mechanics * m, std::vector<Bonus> & converted) const
+{
+
+	for(const auto & b : bonus)
+	{
+		Bonus nb(*b);
+
+		//Moat battlefield effect is always permanent
+		nb.duration = Bonus::ONE_BATTLE;
+
+		if(m->battle()->battleGetDefendedTown() && m->battle()->battleGetSiegeLevel() >= CGTownInstance::CITADEL)
+		{
+			nb.sid = Bonus::getSid32(m->battle()->battleGetDefendedTown()->town->faction->getIndex(), BuildingID::CITADEL);
+			nb.source = Bonus::TOWN_STRUCTURE;
+		}
+		else 
+		{
+			nb.sid = m->getSpellIndex(); //for all
+			nb.source = Bonus::SPELL_EFFECT;//for all
+		}
+		std::set<BattleHex> flatMoatHexes;
+
+		for(const auto & moatPatch : moatHexes)
+			flatMoatHexes.insert(moatPatch.begin(), moatPatch.end());
+
+		nb.limiter = std::make_shared<UnitOnHexLimiter>(std::move(flatMoatHexes));
+		converted.push_back(nb);
+	}
+}
+
+void MoatPlacer::apply(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
+{
+	assert(m->isMassive());
+	assert(m->battle()->battleGetDefendedTown());
+	if(m->isMassive() && m->battle()->battleGetSiegeLevel() >= CGTownInstance::CITADEL)
+	{
+		EffectTarget moat;
+		placeObstacles(server, m, moat);
+
+		std::vector<Bonus> converted;
+		convertBonus(m, converted);
+		for(auto & b : converted)
+		{
+			GiveBonus gb(GiveBonus::EType::BATTLE);
+			gb.bonus = b;
+			server->apply(&gb);
+		}
+	}
+}
+
+void MoatPlacer::placeObstacles(ServerCallback * server, const Mechanics * m, const EffectTarget & target) const
+{
+	assert(m->battle()->battleGetDefendedTown());
+	assert(m->casterSide == BattleSide::DEFENDER); // Moats are always cast by defender
+	assert(turnsRemaining < 0); // Moats should lasts infininte number of turns
+
+	BattleObstaclesChanged pack;
+
+	auto all = m->battle()->battleGetAllObstacles(BattlePerspective::ALL_KNOWING);
+
+	int obstacleIdToGive = 1;
+	for(auto & one : all)
+		if(one->uniqueID >= obstacleIdToGive)
+			obstacleIdToGive = one->uniqueID + 1;
+
+	for(const auto & destination : moatHexes)  //Moat hexes can be different obstacles
+	{
+		SpellCreatedObstacle obstacle;
```

Comment:
Perhaps obstacle (at least partially) should be loaded from config?

Reply:
@IvanSavenko what are you mean? Can you explain more?

Reply:
@IvanSavenko there is only 3 static properties there. Notable:
```c++
		obstacle.casterSide = BattleSide::DEFENDER; // Moats are always cast by defender
		obstacle.passable = true; //Moats always passable
		obstacle.nativeVisible = false; //Moats is invisible for native terrain
```

All other is set by config, yes. Almost half of properties is shared for each obstacle, but I do not know how effect instances work for each spell and for each caster. Is there an object instance for each such pair - your suggestion possible. If not and effect loads from config each time - in theory then we can place obstacles with foreign properties.

Reply:
@IvanSavenko you think than creating template obstacle is better than setting properties from effect always?

Reply:
Template obstacle that *will be directly read from json*
Currently you have (effectively) duplicated code:
1) On loading - code that reads json into Moat class members
2) On casting - Code that copies Moat members into Obstacle members

If it is possible (and makes sense logically) to set it up in a way that:
1) On loading - code that reads json into Obstacle that will be stored as member of Moat. This code can be moved to Obstacle class so Obstacle will be responsible for own state.
2) On casting - code that copies existing obstacle and modifies only necessary fields (e.g. position)

That being said, not insisting on this change.

Reply:
@IvanSavenko in theory, I can use serializeJson method of CObstacleInstance to get a template obstacle, but then we will break mod compatibility (because any obstacle placing spell should be rewritten to use template obstacle, and not use effect settings).

Reply:
@IvanSavenko this is ObstacleSideOptions, but used as obstacle. I think maybe someday we will extend moat class to attackers, and then we also will need side options (now ObstacleSideOptions is used only to store animation).

Reply:
To not to become https://github.com/vcmi/vcmi/pull/363/files, I prefer not to completely redesign obstacle class. Maybe later, but not in this PR.